### PR TITLE
Use large enough buffer for signature in dgst.c

### LINKED ISSUE #9732: Unable to use custom digital signature algorithm with openssl dgst command
<!--
Thank you for your bug report.

NOTE:

    If you're asking about how to use OpenSSL, this isn't the right 
    forum.  Please see our User Support resources:
    https://github.com/openssl/openssl/blob/master/.github/SUPPORT.md

Please remember to tell us in what OpenSSL version you found the issue.

For build issues:

    If this is a build issue, please include the configuration output
    as well as a log of all errors.  Don't forget to include the exact
    commands you typed.

    With OpenSSL before 1.1.1, the configuration output comes from the
    configuration command.  With OpenSSL 1.1.1 and on, it's the output
    of `perl configdata.pm --dump`

For other issues:

    If it isn't a build issue, example code or commands to reproduce
    the issue is highly appreciated.
    Also, please remember to tell us if you worked with your own
    OpenSSL build or if it is system provided.

Please remember to put ``` lines before and after any commands plus
output and code, like this:

    ```
    $ echo output output output
    output output output
    ```

    ```
    #include <stdio.h>
    
    int main() {
        int foo = 1;
        printf("%d\n", foo);
    }
    ```
-->

For the following command line:

```
echo 'Hello world' > sign.txt
openssl dgst -engine <engine_name> -sha256 -sign CA.key -out sign.txt.sha256 sign.txt
```

I got the following output:

```
Error Signing Data
139681221683008:error:0608C09B:digital envelope routines:EVP_PKEY_sign:buffer too small:crypto/evp/pmeth_fn.c:65
```
Used signature schemes signature size is 49216 bytes (and EVP_PKEY_size returns this size).

After debugging session I discovered, that siglen is hardcoded at: `apps/dgst.c:453`

```c
if (key != NULL) {                                                                                                                                                                                                                                
          EVP_MD_CTX *ctx;                                                                                                                                                                                                                              
          BIO_get_md_ctx(bp, &ctx);                                                                                                                                                                                                                     
          len = BUFSIZE;  // <----- HARDCODE                                                                                                                                                                                                                               
          if (!EVP_DigestSignFinal(ctx, buf, &len)) {                                                                                                                                                                                                   
              BIO_printf(bio_err, "Error Signing Data\n");                                                                                                                                                                                              
              ERR_print_errors(bio_err);                                                                                                                                                                                                                
              return 1;                                                                                                                                                                                                                                 
          }                                                                                                                                                                                                                                             
      }
```

Is that an issue of my engines implementation? What should exactly return `EVP_PKEY_size` signature size or maximum digest size?

### PR DESCRIPTION
**Pull request by nblka**
Fixes #9732

### CODE REVIEW

File: apps/dgst.c [hunk 1]
```diff
@@ -523,28 +524,35 @@ int do_fp(BIO *out, unsigned char *buf, BIO *bp, int sep, int binout,
...
+        ret = 0;
+        goto end;
     }
     if (key != NULL) {
         EVP_MD_CTX *ctx;
+        int pkey_len;
         BIO_get_md_ctx(bp, &ctx);
-        len = BUFSIZE;
+        pkey_len = EVP_PKEY_size(key);
+        if (pkey_len > BUFSIZE) {
```
Comment (mattcaswell) [inline apps/dgst.c]: Do we actually have any signature schemes that exceed this limit? Or have you hit this with an engine supplied signature scheme?

File: apps/dgst.c [hunk 2]
```diff
@@ -579,5 +587,11 @@ int do_fp(BIO *out, unsigned char *buf, BIO *bp, int sep, int binout,
         }
         BIO_printf(out, "\n");
     }
-    return 0;
+
+    ret = 0;
+ end:
+    if (NULL != sigbuf) {
```
Comment (mattcaswell) [inline apps/dgst.c]: Normally, we tend to do this the other way around, i.e. `sigbuf != NULL`

File: apps/dgst.c [hunk 3]
```diff
@@ -579,5 +587,11 @@ int do_fp(BIO *out, unsigned char *buf, BIO *bp, int sep, int binout,
...
+        OPENSSL_clear_free(sigbuf, len);
+    }
```
Comment (mattcaswell) [inline apps/dgst.c]: Nit: Don't use {} for a single line if.

**Comment by mattcaswell** [comment]
Close/reopen to kick cla bot

File: apps/dgst.c (hunk 2 above)
Comment (mkwork) [reply apps/dgst.c]: Such practice could avoid misspelling:
e.g:

```c
int a = 42;

if (a = 3) 
   abort(); // you would always abort

if (3 == a) // you would got compilation error here
   abort();
```

Comment (p-steuer) [reply apps/dgst.c]: on the other hand its more difficult to read (at least for lt gt) and afaik most modern compilers would emit a warning telling you to put additional braces around the assignment to indicate that is really what you want to do.

File: apps/dgst.c (hunk 1 above)
Comment (nblka) [reply apps/dgst.c]: This PR is to support our custom engine based signature scheme with large signature sizes. It is originated from #9732. As you commented this issue, PRs are welcome, so I prepared this PR. Hope you'll find it useful. I tried to change existing behavior as little as possible. The new buffer will not be allocated if the existing buffer passed from `dgst_main` is enough to store.

File: apps/dgst.c (hunk 2 above)
Comment (nblka) [reply apps/dgst.c]: Thank you, I've updated this line

File: apps/dgst.c (hunk 3 above)
Comment (nblka) [reply apps/dgst.c]: Got it, fixed

File: apps/dgst.c (hunk 1 above)
Comment (mattcaswell) [reply apps/dgst.c]: No. This is fine as is. I added the labels for the target branches.

**Comment by p-steuer** [comment]
The commit message reads:
``` 
Allocate and use new signature buffer with desired length if existing…

… buffer is not large enough

Fixes #9732
```
Since this description is very general, would you ( @nblka ) mind if it was changed to something along the lines of
```
apps/dgst.c: allocate a new signature buffer

... if the fixed-size buffer is too small. 

Fixes #9732
```
when squashed ?

**Comment by nblka** [comment]
@p-steuer no, feel free to change it

**Comment by p-steuer** [comment]
merged, thanks.

master: https://github.com/openssl/openssl/commit/7c2d95d47ccb3797f0da6bd4446747c6eee07b87
1.1.1: https://github.com/openssl/openssl/commit/43a8f91f00e35b20e33c393f5b79215c277c508e

**Pull request closed without merging**